### ISSUE TITLE
bug(material/chip): Mdc Chip Grid empties all options when using Reactive Forms and focusing out

### ISSUE THREAD

**Issue by vmuresanu** [angular/components#26358]
### Is this a regression?

- [X] Yes, this behavior used to work in the previous version

### The previous version in which this bug was not present was

14.2.6

### Description

Having Reactive forms attached to `mat-chip-grid` which accepts an array of object values, clears them on focus out.

### Reproduction

Steps to reproduce:
1. Create a formgroup with some formControl
2. Bind it to `<mat-chip-grid> `
3. Add some new options (or the same works if the list is already populated) that are **arrays of objects**.
4. Focus out of the component

https://stackblitz.com/edit/angular-6g9jz1?file=src%2Fapp%2Fchips-input-example.ts,src%2Fapp%2Fchips-input-example.html

Upon further investigation, figured that the core problem is that on focus out, object values are transformed to strings hence why we don't see the name property. Nevertheless objects were not transformed in v 14. 

![mat-grid-bug](https://user-images.githubusercontent.com/26407402/210384868-edfda991-132a-48dd-9a44-90aeaeebaebc.gif)

_Note:_ In the Docs Chip list works as expected because it is not bound to Reactive forms

### Expected Behavior

All options remain with previously defined object values

### Actual Behavior

All options are transformed to strings thus we have empty text options

### Environment

- Angular: 15.0.4
- CDK/Material: 15.0.3
- Browser(s): Chrome 108.0.5359.124
- Operating System: macOS Ventura 13.1

**Comment by UrzaUrza** [comment]
Hello,
I'm unable to use the new chips because of this issue.
I can see that the related fix is stuck since the 10 january, is it planned to be fixed soon ?
Are there some workaround ? (Except using the legacy chips)

**Comment by vmuresanu** [comment]
> To clarify - the focus-out event doesn't **clear** the value, it converts it to an array of strings.
> 
> 
> 
> So if we use items as objects like this: `[{ id: 1, name: 'orange'}, { id: 2, name: 'apple'}]`
> 
> then after the focus-out event, the component value becomes `['orange', 'apple]`.
> 
> 
> 
> A very hacky workaround would be to listen to blur event on the <input> element
> 
> `<input ... (blur)="onBlur()">`
> 
> And restore the array of objects by searching the list by the names we got in the component's value...
> 
> 
> 
> 
> 
> ```
> 
> onBlur(event: FocusEvent) {
> 
>      setTimeout(() => this.restoreArrayOfObjects, 0);
> 
> }
> 
>     
> 
> private restoreArrayOfObjects() {
> 
>      // find the objects by their names from formControl.value/ngModel/whatever
> 
>      // and update the value to an array of objects by this.formControl.setValue()
> 
> }
> 
> ```

Agree, this solution is hacky. In fact I made a PR that fixes the issue. There are a couple of solutions how we can do it.
Either by adding a new ’@Input’ that specifies the exact property that needs to be accessed. Or there has to be made a refactor in the edit mode logic. Either way I need input from the code owner since it was 5 months and still no answer.

**Comment by poveu** [comment]
I'll tell you something, the issue seems much bigger than that, as I have a case where your fix won't help.

In some part of the application I'm using chips that display their name + description
```
<mat-chip-row
    *ngFor="let item of selectedItems"
    (removed)="remove(item)">
    <span class="chip-label">{{ item.name }}</span>
    <span class="chip-description">{{ item.description }}</span>
    <mat-icon matChipRemove>cancel</mat-icon>
</mat-chip-row>
```
![image](https://github.com/angular/components/assets/32791888/7bc17244-080e-46c1-90b1-db9d6c8a2a05)

As I have described previously, the focus out (blur) event causes the selected items array to become an array of strings.
However, the bigger issue is that if I'm displaying name + description, then these two values get combined in the array, so the array (ngModel or control's value) becomes:
`['Google Chrome: Fast & Securecom.android.chrome']`

Looks like the focus out event takes the innerText of the chip element and uses it as the item's value, which seems like a huge misconception.

**Comment by vmuresanu** [comment]
@poveu but how does your chips get created? You have only one input to add. So how are you able to define both name and description?

**Comment by poveu** [comment]
As I said before, in my case, my form control (or ngModel) value (currently selected items) is not an array of strings (with just item names) but it's an array of whole objects - so I can display whatever data I want, not just the name.
Look at my modified example:

https://stackblitz.com/edit/ovb2sn?file=src%2Fapp%2Fchips-autocomplete-example.ts

The thing is, even for this example to work, I had to switch from `event.option.viewValue` (which was used by Angular team in their basic example and which is just a string) to `event.option.value` (which is the real value = full object).

**Comment by AlexanderLeonov** [comment]
Ok, so I ran into the same issue as OP and luckily it looks like I was able to figure it out.

There's a mistake in the template in the attached [stackblitz](https://stackblitz.com/edit/ovb2sn?file=src%2Fapp%2Fchips-autocomplete-example.ts) example (the one from the starting post) which causes this issue (see picture below). `[value]` should be bound to `option` instead of `option.name`. This causes re-evaluation of the array on blur which boils down to [here](https://github.com/angular/components/blob/16.2.x/src/material/chips/chip.ts#L201) - which ultimately causes array of objects to turn into an array of strings.

So, unless I missed something else, looks like there's no bug here, just a little misunderstanding.

![image](https://github.com/angular/components/assets/7725399/e0afa013-506b-4e6b-b44c-59d23519d534)

**Comment by mcost45** [comment]
Have the exact same issue.... why was this behaviour changed?

**Comment by poveu** [comment]
I have noticed another issue using formControl, caused by Angular Material's team using `viewValue`.

When inside `<mat-chip-row>` there is any HTML element, or even just the delete button is **wrapped inside some** `<ng-container>`, then whole chip-row's viewValue contains this element's (button with `<mat-icon>cancel</mat-icon>` in this case) innerText...
So for example, `["Apple"]` becomes `["Apple cancel"]`.

Look at my simple example:
https://stackblitz.com/edit/ovb2sn-s41kjs?file=src%2Fapp%2Fchips-autocomplete-example.html,src%2Fapp%2Fchips-autocomplete-example.ts

All you have to do to reproduce the issue is to click on the text input and then focus it out. This will run Material's blur/focusOut code and update the form control value with the chips' viewValue which becomes `["Apple cancel"]`.

**Comment by jonastg** [comment]
Very good point for @AlexanderLeonov !

> Ok, so I ran into the same issue as OP and luckily it looks like I was able to figure it out.
> 
> There's a mistake in the template in the attached [stackblitz](https://stackblitz.com/edit/ovb2sn?file=src%2Fapp%2Fchips-autocomplete-example.ts) example (the one from the starting post - just copying link here for convenience) which causes this issue (see picture below). `[value]` should be bound to `option` instead of `option.name`. Losing focus causes re-evaluation of the array which boils down to [here](https://github.com/angular/components/blob/16.2.x/src/material/chips/chip.ts#L201) - which ultimately causes array of objects to turn into an array of strings.
> 
> So, unless I missed something else, looks like there's no bug here, just a little misunderstanding.
> 
> ![image](https://user-images.githubusercontent.com/7725399/266725218-e0afa013-506b-4e6b-b44c-59d23519d534.png)

Is it a bug or a feature? 

I've checked out the documentation for Chips and [mat-chip-row](https://v15.material.angular.io/components/chips/api#MatChipRow) `@Input() value: any | The value of the chip. Defaults to the content inside the mat-mdc-chip-action-label element.`. It specifies that will take the label content if there are no value. I think it is a big change that it isn't reflected in examples and it pass by unnoticed.

**Comment by poveu** [comment]
That is very true @AlexanderLeonov & @jonastg !
Setting `[value]` param for `mat-chip-row` even fixes the issue I have reported a few posts above.

**Comment by AlexanderLeonov** [comment]
In my opinion, just defaulting to innerText is not a very good idea, and this thread is a very good example of why, especially the last issue reported by @poveu. I think this is not a bug in the sense that the code behaves as designed and doesn't just break unexpectedly. The issue here is that the design itself is very confusing and unclear. I think the component should kick and scream at you if you don't provide a value the way it expects it to be provided. Silent crippling the data without any explanations doesn't seem acceptable to me.

**Comment by poveu** [comment]
...the data that makes no sense, like the "cancel" part of the removal button.
I agree 100%.

**Comment by vmuresanu** [comment]
I think that since `[value]` appears to be from logical perspective, mandatory for `<mat-chip-row>`, and since we already have angular 16 in material, we can benefit from required inputs. The solution would be to make @Input({requred: true}) value in MatChip source code. Also would be great if this information was documented. So I guess there are a couple of action points to be done.

**Comment by vmuresanu** [comment]
I would also like to warn, that by changing `[value]="option"` from `[value]="option.name"` when enabling chip edit we get `[object Object]` in the input, because `MatChipEditInput` is getting the `textContent` from the input, which is our case is an object.
I have refactored the original example by removing unnecessary code and making add / edit / delete of a chip working with reactive forms approach and with `[value]="option"` - https://stackblitz.com/edit/angular-6g9jz1-zkc8ak?file=src%2Fapp%2Fchips-input-example.ts,src%2Fapp%2Fchips-input-example.html

**Comment by AlexanderLeonov** [comment]
Yeah... That's an issue too albeit a different one. Mat components can't really make any assumptions about the object being bound to the value property, hence they should ask app to provide some info/mappings/whatever it is about those objects, or they should clearly describe what assumptions mat components impose on those objects (like having certain field or something) in order for other functionality - like tag editing - to work correctly. This looks either like a missing functionality or a design flaw.

I suspect you can work around this particular issue by redefining `.toString()` in the tag objects but it feels more like a hack rather than a solution.

**Comment by liesahead** [comment]
The issue is when using chips with autocomplete that you can't use some `compareWith` function (like with mat-select), so if you're patching values from one array of objects and then going through another array where references are not the same there's no way to compare objects by ids.

It's feels like this wasn't designed with objects in mind, just primitives..

**Comment by xsegrity** [comment]
> Very good point for [@AlexanderLeonov](https://github.com/AlexanderLeonov) !
> 
> > Ok, so I ran into the same issue as OP and luckily it looks like I was able to figure it out.
> > There's a mistake in the template in the attached [stackblitz](https://stackblitz.com/edit/ovb2sn?file=src%2Fapp%2Fchips-autocomplete-example.ts) example (the one from the starting post - just copying link here for convenience) which causes this issue (see picture below). `[value]` should be bound to `option` instead of `option.name`. Losing focus causes re-evaluation of the array which boils down to [here](https://github.com/angular/components/blob/16.2.x/src/material/chips/chip.ts#L201) - which ultimately causes array of objects to turn into an array of strings.
> > So, unless I missed something else, looks like there's no bug here, just a little misunderstanding.
> > ![image](https://user-images.githubusercontent.com/7725399/266725218-e0afa013-506b-4e6b-b44c-59d23519d534.png)
> 
> Is it a bug or a feature?
> 
> I've checked out the documentation for Chips and [mat-chip-row](https://v15.material.angular.io/components/chips/api#MatChipRow) `@Input() value: any | The value of the chip. Defaults to the content inside the mat-mdc-chip-action-label element.`. It specifies that will take the label content if there are no value. I think it is a big change that it isn't reflected in examples and it pass by unnoticed.

This is the way!!! Thank you. I can finally rest my head after all the wall banging.
